Storage target mode, inspired by "target disk mode" á la MacOS, but with NVMe-TCP

poettering opened the pull request:
This implements a "storage target mode", similar to what MacOS provides
since a long time:

→ https://en.wikipedia.org/wiki/Target_Disk_Mode

This implementation is relatively simple:

1. a new generic target "tgtmode.target" is added, which when booted
   into defines the target mode.

2. a small tool and service "systemd-tgtmode" is added which exposes a
   specific device or all devices as NVMe-TCP devices over the network.
   NVMe-TCP appears to be hot shit right now how to expose block devices
   over the network. And it's really simple to set up via configs, hence
   our code is relatively short and neat.

The idea is that tgtmode.target can be extended sooner or later, for
example to expose block devices also as USB mass storage devices and
similar, in case the system has "dual mode" USB controller that can also
work as device, not just as host.

How to use this? Boot into your system with a kernel cmdline of
"rd.systemd.unit=tgtmode.target ip=dhcp", and you'll see on screen the
precise "nvme connect" command line to make the relevant block devices
available locally on some other machine. This all requires that the
target mode stuff is included in the initrd of course. And the system
will the stay in the initrd forever.

Why bother? Primarily two use-case:

1. Debug a broken system: with very few dependencies during boot get
   access to the raw block device of a broken machine.

2. Migrate from system to another system, by dd'ing the old to the new
   directly.

(And there might be more, for example the ability to boot from a
laptop's disk on another system)

Limitations:

1. There's no authentication/encryption. Hence: use this on local links
   only.

2. NVMe target mode on Linux supports r/w operation only. Ideally, we'd
   have a read-only mode, for security reasons, and default to it.

Future love:

1. We probably should set up a Thunderbolt networking device
   automatically if two laptops are connected directly via thunderbolt.

2. We should have another mode, where we simply expose the homed LUKS
   home dirs like that.

3. For security reasons maybe add ip=link-local or so as new setting for
   systemd-network-generator that exclusively configures IPv4LL + IPv6LL
   but no public addresses, so that the system is only reachable on the
   local link.

poettering commented:
i am not a fan of the name "tgtmode" for this, but other things such as "targetdm" sounded weird to me as you then boot into targetdm.target and that's a lot of target in one word.

other ideas welcome? tgtdsk? tgtdm? target-disk-mode?

Winterhuman commented:
What about:

- `netdisk.target`
- `netnvme.target`
- `tcpdisk.target`
- `nvmeshare.target`

poettering commented:
Well, the target is supposed to be named somewhat generically, independent of the protocol/transport used. I.e. i think for devices that have a dual-mode usb controller (raspberry pis for example) it would make a ton of sense to expose the disk as usb mass storage to hosts. Hence "nvme" or "tcp" should not be in the name

Winterhuman commented:
Ah, got it. I will say that it isn't immediately obvious what the "target" in "target disk mode" is supposed to mean; a new user seeing the term "target disk mode" probably won't understand what the function of it is, so including `net`, `network`, or `remote` in the name would help

dnicolodi commented:
What's the problem in calling this `target-disc-mode`? I don't think keeping the name short is important.

rpigott commented:
I don't hate tgtmode, but if you're looking for more names maybe something with "fabric" in the name would be pleasant? nvme-over-fabrics seems to be the common term to refer to the whole class of nvme transports in the network storage world.

poettering commented:
So I renamed the service to `systemd-storagetm.service` now, and the target to `storage-target-mode.target`. 

I also added docs, should be ready to review.

poettering commented:
/cc @hreinecke hey, maybe this is interesting to you, given you work on nvme iirc?

bluca commented on man/systemd-storagetm.service.xml:
```diff
@@ -0,0 +1,99 @@
+    functionality should hence only be used in a local setup.</para>
+
+    <para>Note that to function properly networking must be configured too. The recommended mechanism to boot
+    into a storage target mode is by adding <literal>rd.systemd.unit=storage-target-mode.target
```
Can this be done with a credential? If so, it should be documented too as an example

poettering replied:
we currently do not support that in systemd, no, sorry.

poettering commented:
> 3\. For security reasons maybe add ip=link-local or so as new setting for
>         systemd-network-generator that exclusively configures IPv4LL + IPv6LL
>         but no public addresses, so that the system is only reachable on the
>         local link.

THis is implemented now in https://github.com/systemd/systemd/pull/29769

lf- commented:
> So I renamed the service to `systemd-storagetm.service` now, and the target to `storage-target-mode.target`.
> 
> I also added docs, should be ready to review.

I think still `systemd-storagetm.service` is not necessarily obvious it is related to `storage-target-mode.target`. Maybe just eat the keystrokes and call it `systemd-storage-target-mode`, since it doesn't feel like something people would be typing so much.

poettering commented:
> I think still `systemd-storagetm.service` is not necessarily obvious it is related to `storage-target-mode.target`. Maybe just eat the keystrokes and call it `systemd-storage-target-mode.service`, since it doesn't feel like something people would be typing so much.

I actually chose a different name on purpose, because one is supposed to be generic, that people can fill with meaning as they like and the other one is a specific implementation.

For example, for specific devices it might make sense to also expose things as NBD, iSCSI, USB mass storage, whatever you want, and people should be able to just hook this into the target too, to get both implementations activated at the same time. Ours that does NVMe-TCP and something else.

hence:

1. storage-target-mode.target → the generic concept
2. systemd-storagetm.service → one possible implementation of it, among theoretically many, that deals with NVMe-TCP

flokli commented:
We have `remote-fs.target`, to sync on filesystem mounts needing network access, what about calling this one `remote-block.target`?

Not sure about `systemd-storagetm.service` though - if it's the official systemd daemon to mount NVMe-TCP devices, it being called `systemd-nvme-tcpd.service` would feel more intuitive.

hreinecke commented:
Hmm. Have you considered using 'nvmetcli' to configure the target?
Or, alternatively: can you parse the json files from nvmetcli for this service?
That would be awesome, as then one could pre-configure the target, and just use this service to start it up.
But maybe I'm misunderstanding the whole concept, seeing that last time I worked on MacOS was with 7.0.1 ...

hreinecke commented:
And it's ever so slightly incorrect that NVMe does not support authentication nor encryption. It does support authentication since 6.0. And encryption is queued for 6.7.

hreinecke commented:
After a cursory glance for the service, how do you identify which network interfaces to use?
I do assume that this service is started after the network service, right?
And I further assume that one needs to figure out the configured network interface only, and not 'all' interfaces, right?

poettering commented:
> We have `remote-fs.target`, to sync on filesystem mounts needing network access, what about calling this one `remote-block.target`?

the word "remote" implies networked storage I'd claim. But this concept is supposed to eventually also expose USB mass storage where that can be done.

> Not sure about `systemd-storagetm.service` though - if it's the official systemd daemon to mount NVMe-TCP devices, it being called `systemd-nvme-tcpd.service` would feel more intuitive.

nope. it doesn't *mount* devices, it exposes local block devices over NVMe-TCP for other hosts so that they can mount them.

Also, I don't think we should encode the transport in the name, because there's also NVMe-FC, and other stuff, i.e. and while I don't see right now why it would make sense for this little tool to support that, we should keep the door open to add other transports in future.

Also, it might make sense to add USB mass storage exposure stuff directly to the tool, since it uses very similar kernel apis via configfs these days, and hence would share most behaviour anyway.

poettering commented:
> Hmm. Have you considered using 'nvmetcli' to configure the target?
> Or, alternatively: can you parse the json files from nvmetcli for this service?

I looked at it, but I want to do this form C, not shell out to something else. Also AFAICS nvmetcli is basically a 1:1 wrapper of configs writes, which we can do natively easier, without adding a dep.

> That would be awesome, as then one could pre-configure the target, and just use this service to start it up.

Well, they key of this stuff here is that it's entirely automatic: you boot into the mode and don't configure a thing, it just exposes everything.

I think it would make a ton of sense for nvmetcli to come with its own .service file, but this is quite a different usecase than this, because it implies configuration and stuff, which we really don#t want here.

> And it's ever so slightly incorrect that NVMe does not support authentication nor encryption. It does support authentication since 6.0. And encryption is queued for 6.7.

It does for the target? how would i configure that? are there any docs how to configure this on the target?

> After a cursory glance for the service, how do you identify which network interfaces to use?

The idea is that this is combined with `ip=dhcp` on the kernel cmdline or something similar which tells `systemd-networkd` to configure IP on anything thta vaguely looks like ethernet. The NVMe-TCP is then configured to listen on `0.0.0.0`/`::`, i.e. accept traffic from everywhere.

> I do assume that this service is started after the network service, right?

No it's not. Why would it? Since we do a wildcard IP address bind it shouldn't matter.

(We do subscribe to rtnetlink though to get notifications whenever the IP config changes, so that we can display on screen where to connect to)

> And I further assume that one needs to figure out the configured network interface only, and not 'all' interfaces, right?

We simply show the relevant public IP addresses on screen. Basically, if you boot up on this mode you'll see something like this on screen:

```
NVMe-TCP: → 🌍 Listening on ipv4 (port 42316)
          ├─ Try for all devices: nvme connect-all -t tcp -a 192.168.100.1 -s 42316
          └─ Try for all devices: nvme connect-all -t tcp -a 172.31.0.46 -s 42316
NVMe-TCP: → 🌍 Listening on ipv6 (port 34475)
          ├─ Try for all devices: nvme connect-all -t tcp -a fd00::ae0e:f19d:56fc:e003 -s 34475
          └─ Try for all devices: nvme connect-all -t tcp -a fe80::6043:4e3a:8703:e2fd -s 34475
NVMe-TCP: → 💽 nqn.2023-10.io.systemd:storagetm.2d4c76c2cb1d4866b26930eabaf108d8.nullb0 (/dev/nullb0)
          ├─ Try for specific device: nvme connect -t tcp -n 'nqn.2023-10.io.systemd:storagetm.2d4c76c2cb1d4866b26930eabaf108d8.nullb0' -a 192.168.100.1 -s 42316
          ├─ Try for specific device: nvme connect -t tcp -n 'nqn.2023-10.io.systemd:storagetm.2d4c76c2cb1d4866b26930eabaf108d8.nullb0' -a 172.31.0.46 -s 42316
          ├─ Try for specific device: nvme connect -t tcp -n 'nqn.2023-10.io.systemd:storagetm.2d4c76c2cb1d4866b26930eabaf108d8.nullb0' -a fd00::ae0e:f19d:56fc:e003 -s 34475
          └─ Try for specific device: nvme connect -t tcp -n 'nqn.2023-10.io.systemd:storagetm.2d4c76c2cb1d4866b26930eabaf108d8.nullb0' -a fe80::6043:4e3a:8703:e2fd -s 34475
NVMe-TCP: → 💽 nqn.2023-10.io.systemd:storagetm.2d4c76c2cb1d4866b26930eabaf108d8.dm-2 (/dev/dm-2)
          ├─ Try for specific device: nvme connect -t tcp -n 'nqn.2023-10.io.systemd:storagetm.2d4c76c2cb1d4866b26930eabaf108d8.dm-2' -a 192.168.100.1 -s 42316
          ├─ Try for specific device: nvme connect -t tcp -n 'nqn.2023-10.io.systemd:storagetm.2d4c76c2cb1d4866b26930eabaf108d8.dm-2' -a 172.31.0.46 -s 42316
          ├─ Try for specific device: nvme connect -t tcp -n 'nqn.2023-10.io.systemd:storagetm.2d4c76c2cb1d4866b26930eabaf108d8.dm-2' -a fd00::ae0e:f19d:56fc:e003 -s 34475
          └─ Try for specific device: nvme connect -t tcp -n 'nqn.2023-10.io.systemd:storagetm.2d4c76c2cb1d4866b26930eabaf108d8.dm-2' -a fe80::6043:4e3a:8703:e2fd -s 34475
```

```

yuwata commented on src/storagetm/storagetm.c:
```diff
@@ -0,0 +1,1064 @@
+        assert(argc >= 0);
+        assert(argv);
+
+        while ((c = getopt_long(argc, argv, "h", options, NULL)) >= 0) {
```
Missing 'a'

yuwata commented on src/storagetm/storagetm.c:
```diff
@@ -0,0 +1,1064 @@
+
+        if (arg_all > 0) {
+                if (argc > optind)
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Expects no further arguments if --all is specified.");
```
Maybe better to mention the short option too.

yuwata commented on src/storagetm/storagetm.c:
```diff
@@ -0,0 +1,1064 @@
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Expects no further arguments if --all is specified.");
+        } else {
+                if (optind >= argc)
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Expecting device name or --all.");
```
Also here.

yuwata commented on src/storagetm/storagetm.c:
```diff
@@ -0,0 +1,1064 @@
+        if (!linkname)
+                return log_oom();
+
+        if (unlinkat(port->nvme_port_fd, linkname, 0) < 0)
```
Maybe better to ignore ENOENT?

yuwata commented on src/storagetm/storagetm.c:
```diff
@@ -0,0 +1,1064 @@
+static void context_done(Context *c) {
+        assert(c);
+
+        c->ipv4_port = nvme_port_destroy(c->ipv4_port);
```
Not a matter, but we usually not assign 0 or NULL in xyz_done().

poettering replied:
we sometimes do, and sometimes do not ;-)

yuwata commented on src/storagetm/storagetm.c:
```diff
@@ -0,0 +1,1064 @@
+        assert(d);
+        assert(ret);
+
+        const char *devname;
```
Please initialize with NULL, or make the failure in the getter below critical.

yuwata reviewed (changes requested):
Please use log_device_error_errno() and friends where applicable.

poettering commented on src/storagetm/storagetm.c:
```diff
@@ -0,0 +1,1064 @@
+                        }
+                }
+
+                subsystems_dir_fd = safe_close(subsystems_dir_fd);
```
it's not actually, the idea is that the fd indicates whether we still need to unlink the subdir or not. will add comment.

poettering commented:
new version, with raised points addressed.

AsciiWolf commented on TODO:
```diff
@@ -175,18 +175,18 @@ Features:
-  first ask for the local root pw, to authenticate that logging in like this is
-  ok, and would then be followed by another password prompt asking for the
-  user's own password. Also, do something similar for CIFS: if you log in via
+* homed: use systemd-storagetm to expose expose home dirs via nvme-tcp. Then,
```
```suggestion
* homed: use systemd-storagetm to expose home dirs via nvme-tcp. Then,
```

Small typo?

AsciiWolf commented on man/systemd-storagetm.service.xml:
```diff
@@ -0,0 +1,104 @@
+    <title>Description</title>
+
+    <para><filename>systemd-storagetm.service</filename> is a service that exposes all local block devices as
+    NVMe-TCP mass storage devices. It's primary use-case is to be invoked by the
```
```suggestion
    NVMe-TCP mass storage devices. Its primary use-case is to be invoked by the
```

Small typo.

yuwata commented on src/storagetm/storagetm.c:
```diff
@@ -0,0 +1,1058 @@
+        _cleanup_free_ char *link = NULL;
+        int r;
+
+        r = terminal_urlify_man("systemd-storagetm", "1", &link);
```
`1` -> `8`

yuwata commented on src/storagetm/storagetm.c:
```diff
@@ -0,0 +1,1058 @@
+static int parse_argv(int argc, char *argv[]) {
+
+        enum {
+                ARG_READ_ONLY = 0x100,
```
Unused ?

yuwata commented on src/storagetm/storagetm.c:
```diff
@@ -0,0 +1,1058 @@
+        assert(argc >= 0);
+        assert(argv);
+
+        while ((c = getopt_long(argc, argv, "ha", options, NULL)) >= 0) {
```
Not a matter, but `{` is not necessary.

yuwata commented on src/storagetm/storagetm.c:
```diff
@@ -0,0 +1,1058 @@
+                                        if (r < 0)
+                                                log_warning_errno(r, "Failed to disable namespace '%s' of NVME subsystem '%s', ignoring: %m", e->d_name, s->name);
+
+                                        if (unlinkat(namespaces_fd, e->d_name, AT_REMOVEDIR) < 0)
```
This may be racy, better to ignore ENOENT?

yuwata commented on src/storagetm/storagetm.c:
```diff
@@ -0,0 +1,1058 @@
+                        r = readdir_all(namespaces_fd, RECURSE_DIR_SORT|RECURSE_DIR_IGNORE_DOT, &de);
+                        if (r < 0)
+                                log_warning_errno(r, "Failed to read 'namspaces' dir of subsystem '%s', ignoring: %m", s->name);
+                        else {
```
Not a matter, but `{` is not necessary.

yuwata commented on src/storagetm/storagetm.c:
```diff
@@ -0,0 +1,1058 @@
+        }
+
+        if (s->nvme_subsystems_fd >= 0 && s->name) {
+                if (unlinkat(s->nvme_subsystems_fd, s->name, AT_REMOVEDIR) < 0)
```
Better to ignore `ENOENT` ??

yuwata commented on src/storagetm/storagetm.c:
```diff
@@ -0,0 +1,1058 @@
+                        }
+                }
+
+                subsystems_dir_fd = safe_close(subsystems_dir_fd);
```
Maybe better to use `_cleanup_close_` instead?

yuwata commented on src/storagetm/storagetm.c:
```diff
@@ -0,0 +1,1058 @@
+                int ports_fd,
+                uint16_t portnr,
+                int ip_family,
+                int *ret_fd) {
+        int r;
```
We usually insert an empty line.

yuwata commented on src/storagetm/storagetm.c:
```diff
@@ -0,0 +1,1058 @@
+static int nvme_port_add(const char *name, int ip_family, NvmePort **ret) {
+        int r;
+
+        assert(IN_SET(ip_family, AF_INET, AF_INET6));
```
Missing assertion for `name` ?

yuwata commented on src/storagetm/storagetm.c:
```diff
@@ -0,0 +1,1058 @@
+static int nvme_port_report(NvmePort *port) {
+        assert(port);
+
+        if (!port)
```
This will never hit. Maybe, the assertion in the above is redundant??

yuwata replied:
yeah. So, please drop the assertion. If IPv6 is off, the assertion will be triggered.

yuwata commented on src/storagetm/storagetm.c:
```diff
@@ -0,0 +1,1058 @@
+                   af_to_ipv4_ipv6(port->ip_family),
+                   port->portnr);
+
+        FOREACH_ARRAY(a, addresses, n_addresses) {
```
Unnecessary `{`.

yuwata reviewed (approved):
Superficially LGTM, except for https://github.com/systemd/systemd/pull/29748#discussion_r1378599526.

bluca commented:
@mrc0mmand could you please add nvme-cli for archlinux/centos test runs? I've added it for debian/ubuntu

bluca commented:
also please add a NEWS entry

poettering commented:
Force pushed new version with all suggested changes applied.

I generalized the "flock with timeout" thing a bit, because we have very similar code in "udevadm lock". It now uses the same logic, as it should.

PTAL

bluca commented on src/basic/lock-util.c:
```diff
@@ -186,3 +189,85 @@ int lock_generic(int fd, LockType type, int operation) {
+
+        /* A version of lock_generic(), but with a time-out. We do this in a child process, since the kernel
+         * APIs natively don't support a timeout. We set a SIGALRM timer that will kill the child after the
+         * timeout is hit. Returns -ETIMEDOUT if the time-out is hit, and 0 on success. */
```
This still confuses me - it takes the lock in the child process, but then waits for it to exit, so when it returns the lock isn't held anymore no? So it just checks that at some point in time nothing else was holding it? What am I missing?

poettering replied:
BSD locks are per fd. hence the child can take it or the parent, doesn't matter, same effect. after the lock was taken the lock is on the fd, hence all good.

bluca replied:
Ah so it's a way to have the child take the lock instead of the parent, but it is shared. Please expand the comment to say it, so that those of us that are thicker than average don't spend too much time scratching their heads

poettering replied:
added a comment now.

poettering commented:
> 11:13:52 /systemd-meson-build/../root/systemd/src/basic/lock-util.c:225: undefined reference to `timer_create'
> 11:13:52 /systemd-meson-build/../root/systemd/src/basic/lock-util.c:233: undefined reference to `timer_settime'

should be fixed now.

mrc0mmand commented:
> @mrc0mmand could you please add nvme-cli for archlinux/centos test runs? I've added it for debian/ubuntu

Sure, done.

bluca commented:
> > And it's ever so slightly incorrect that NVMe does not support authentication nor encryption. It does support authentication since 6.0. And encryption is queued for 6.7.
> 
> It does for the target? how would i configure that? are there any docs how to configure this on the target?

It was merged for v6.7 it looks like:

https://lwn.net/Articles/942817/

bluca commented:
mmh looks like the test sometimes fails:

```
8100s [   28.414212] testsuite-84.sh[376]: + dd if=/dev/nvme1n1 bs=1024
8100s [   28.531593] testsuite-84.sh[376]: dd: failed to open '/dev/nvme1n1': No such file or directory
8100s [   28.544949] testsuite-84.sh[377]: cmp: EOF on - which is empty
```

mrc0mmand commented:
> mmh looks like the test sometimes fails:
> 
> ```
> 8100s [   28.414212] testsuite-84.sh[376]: + dd if=/dev/nvme1n1 bs=1024
> 8100s [   28.531593] testsuite-84.sh[376]: dd: failed to open '/dev/nvme1n1': No such file or directory
> 8100s [   28.544949] testsuite-84.sh[377]: cmp: EOF on - which is empty
> ```

Yeah, I opened https://github.com/systemd/systemd/pull/29851 which will, hopefully, help with that.
